main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification3.scala

```diff
@@ -0,0 +1,308 @@
+            val tpe2 = fromSetFormula(eq.f2, eq.loc)
+            Result.Err(UnificationError.MismatchedEffects(tpe1, tpe2))
+          case Status.Timeout(msg) =>
+            Result.Err(UnificationError.TooComplex(msg, eq.loc))
+        }
+    }
+  }
+
+  def unify(tpe1: Type, tpe2: Type, scope: Scope, renv: RigidityEnv): Result[Option[Substitution], UnificationError] = {
+    implicit val bimap: Bimap[Atom, Int] = try {
+      mkBidirectionalVarMap(tpe1, tpe2)
+    } catch {
+      case InternalCompilerException(_, _) => return Result.Ok(None)
+    }
+
+    val equation = try {
+      toEquation(tpe1, tpe2, tpe1.loc)(scope, renv, bimap)
+    } catch {
+      case InternalCompilerException(_, _) => return Result.Ok(None)
+    }
+
+    SetUnification.solve(List(equation))(SetUnification.SolverListener.doNothing, Options.default) match {
+      case (Nil, subst) =>
+        Result.Ok(Some(fromSetSubst(subst)))
+
+      case (eq :: _, _) =>
+        // Equations are leftover.
+        eq.status match {
+          case Status.Pending => throw InternalCompilerException(s"Unexpected pending formula $eq", eq.loc)
+          case Status.Unsolvable =>
+            val tpe1 = fromSetFormula(eq.f1, eq.loc)
+            val tpe2 = fromSetFormula(eq.f2, eq.loc)
+            Result.Err(UnificationError.MismatchedEffects(tpe1, tpe2))
+          case Status.Timeout(msg) =>
+            Result.Err(UnificationError.TooComplex(msg, eq.loc))
+        }
+
+    }
+  }
+
+  /**
+    * Returns a bi-directional map from type variables to ints computed from the given list of
+    * unification equations `l`.
+    */
+  private def mkBidirectionalVarMap(l: List[(Type, Type, SourceLocation)]): Bimap[Atom, Int] = {
+    // Find all atoms that occur in anywhere in `l`.
+    val atoms = l.foldLeft(Set.empty[Atom]) {
+      case (acc, (t1, t2, _)) => acc ++ getAtoms(t1) ++ getAtoms(t2)
+    }
+
+    // Construct the map from atoms to ints.
+    mkBidirectionalVarMap(atoms)
+  }
+
+  private def mkBidirectionalVarMap(tpe1: Type, tpe2: Type): Bimap[Atom, Int] = {
+    mkBidirectionalVarMap(getAtoms(tpe1) ++ getAtoms(tpe2))
+  }
+
+
+  private def mkBidirectionalVarMap(atoms: Set[Atom]): Bimap[Atom, Int] = {
+    Bimap.from(atoms.toList.zipWithIndex)
+  }
+
+  /** Returns the set of atoms that occur in the given type. */
+  private def getAtoms(t: Type): Set[Atom] = t match {
+    case Type.Var(sym, _) => Set(Atom.Var(sym))
+    case Type.Cst(TypeConstructor.Effect(sym), _) => Set(Atom.Eff(sym))
+    case Type.Cst(TypeConstructor.Error(id, kind), _) => Set(Atom.Error(id, kind))
+    case Type.Cst(_, _) => Set.empty
+    case Type.Apply(tpe1, tpe2, _) => getAtoms(tpe1) ++ getAtoms(tpe2)
+    case Type.Alias(_, _, tpe, _) => getAtoms(tpe)
+    case Type.AssocType(AssocTypeConstructor(sym, _), arg, kind, _) => Set(Atom.Assoc(sym, getAssocArg(arg), kind))
+    case Type.JvmToEff(_, loc) => throw InternalCompilerException("dont know", loc)
```

Comment:
"don't" :P 

or a slightly better message

Reply:
Unexpected type: .. + add some documentation to unifyAll and unifyOne saying that all `JvmToEff` and `JvmToType` and `UnresolvedJvmType` must have been taken care of. 

Actually, I would assume the latter two can throw ICE with "Unexpected type of illegal kind".

But what about `JvmToEff`? We assume it to be resolved?